src/consensus/tower.zig

```diff
@@ -0,0 +1,3177 @@
+        else
+            null;
+    }
+
+    /// Returns `Some(gca)` where `gca` is the greatest (by slot number)
+    /// common ancestor of both `slot_a` and `slot_b`.
+    ///
+    /// Returns `null` if:
+    /// * `slot_a` is not in `ancestors`
+    /// * `slot_b` is not in `ancestors`
+    /// * There is no common ancestor of slot_a and slot_b in `ancestors`
+    fn greatestCommonAncestor(
+        ancestors: *const AutoHashMapUnmanaged(
+            Slot,
+            SortedSet(Slot),
+        ),
+        slot_a: Slot,
+        slot_b: Slot,
+    ) ?Slot {
+        var ancestors_a = ancestors.get(slot_a) orelse return null;
+        var ancestors_b = ancestors.get(slot_b) orelse return null;
+
+        var max_slot: ?Slot = null;
+
+        for (ancestors_a.items()) |slot| {
+            for (ancestors_b.items()) |other_slot| {
+                if (slot == other_slot and
+                    (max_slot == null or
+                    slot > max_slot.?))
+                {
+                    max_slot = slot;
+                }
+            }
+        }
+
+        return max_slot;
+    }
+
+    // Optimistically skip the stake check if casting a vote would not increase
+    // the lockout at this threshold. This is because if you bounce back to
+    // voting on the main fork after not voting for a while, your latest vote
+    // might pop off a lot of the votes in the tower. The stake from these votes
+    // would have rolled up to earlier votes in the tower, which presumably
+    // could have helped us pass the threshold check. Worst case, we'll just
+    // recheck later without having increased lockouts.
+    fn optimisticallyBypassVoteStakeThresholdCheck(
+        // Needs to be an iterator that produces Lockout
+        tower_before_applying_vote: anytype,
+        threshold_vote: Lockout,
+    ) bool {
+        for (tower_before_applying_vote) |old_vote| {
+            if (old_vote.slot == threshold_vote.slot and
+                old_vote.confirmation_count == threshold_vote.confirmation_count)
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    fn checkVoteStakeThreshold(
+        logger: ScopedLogger(@typeName(Tower)),
+        maybe_threshold_vote: ?Lockout,
+        tower_before_applying_vote: anytype,
+        threshold_depth: usize,
+        threshold_size: f64,
+        slot: Slot,
+        voted_stakes: *const AutoHashMapUnmanaged(Slot, u64),
+        total_stake: u64,
+    ) ThresholdDecision {
+        const threshold_vote = maybe_threshold_vote orelse {
+            // Tower isn't that deep.
+            return ThresholdDecision{ .passed_threshold = {} };
+        };
+
+        const fork_stake = voted_stakes.get(threshold_vote.slot) orelse {
+            // We haven't seen any votes on this fork yet, so no stake
+            return ThresholdDecision{
+                .failed_threshold = .{ @as(u64, threshold_depth), 0 },
```

Comment:
Doesn't seem like there's a need for the `@as`
```suggestion
                .failed_threshold = .{ threshold_depth, 0 },
```

Reply:
[4be8a58](https://github.com/Syndica/sig/pull/664/commits/4be8a583187a899c3bd4374e50ce9ffe7e142662)